From f7d6d433e186a025c2547def4ec6b286ea5f24d6 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 Mar 2021 17:57:27 -0500
Subject: [PATCH] Always require PrecomputedTransactionData in
 GenericTransactionSignatureChecker. (#109)

Basically deletes the constructor which sets txdata to nullptr.

This change is required for BIP341 sighashes, as PrecomputedTransactionData contains m_spent_outputs, which is required for computing parts of the sighash.

Rebase provenance: this historical change was adapted onto the pinned
Bitcoin ABC base. Source diff is the recorded Lotus V1 replay commit below.

Historical-source-commit: d473cab6f4dd3dc27dd70917f84ede7db59c63aa
Recorded-replay-commit: a1d89d83d67234cd770917f902c260bfdb4e9380

diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index 07a1634a93..e39e1daa13 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -698,7 +698,7 @@ static void MutateTxSign(CMutableTransaction &tx, const std::string &flagStr) {
         const Amount amount = coin.GetTxOut().nValue;
 
         SignatureData sigdata =
-            DataFromTransaction(mergedTx, i, coin.GetTxOut());
+            DataFromTransaction(mergedTx, i, txdata);
         // Only sign SIGHASH_SINGLE if there's a corresponding output:
         if ((sigHashType.getBaseType() != BaseSigHashType::SINGLE) ||
             (i < mergedTx.vout.size())) {
diff --git a/src/iguana/iguana_interpreter.h b/src/iguana/iguana_interpreter.h
index 5ac5f05c71..a3d6c0c6ef 100644
--- a/src/iguana/iguana_interpreter.h
+++ b/src/iguana/iguana_interpreter.h
@@ -56,7 +56,8 @@ public:
     IguanaInterpreter(CMutableTransaction txIn, uint32_t inputIndexIn,
                       CTxOut spentOutput, uint32_t flagsIn)
         : tx(txIn), inputIndex(inputIndexIn),
-          scriptPubKey(spentOutput.scriptPubKey), txdata(tx),
+          scriptPubKey(spentOutput.scriptPubKey),
+          txdata(tx, std::vector<CTxOut>(tx.vin.size(), spentOutput)),
           sigChecker(&tx, inputIndex, spentOutput.nValue, txdata),
           flags(flagsIn) {}
 
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index e0decbb2a8..4271e32c33 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -522,7 +522,7 @@ static RPCHelpMan combinerawtransaction() {
                 for (const CMutableTransaction &txv : txVariants) {
                     if (txv.vin.size() > i) {
                         sigdata.MergeSignatureData(
-                            DataFromTransaction(txv, i, txout));
+                            DataFromTransaction(txv, i, txdata));
                     }
                 }
                 ProduceSignature(DUMMY_SIGNING_PROVIDER,
diff --git a/src/script/interpreter.h b/src/script/interpreter.h
index 0d9f3b8683..8220b7f729 100644
--- a/src/script/interpreter.h
+++ b/src/script/interpreter.h
@@ -105,9 +105,6 @@ private:
     mutable SigHashCache m_sighash_cache;
 
 public:
-    GenericTransactionSignatureChecker(const T *txToIn, unsigned int nInIn,
-                                       const Amount &amountIn)
-        : txTo(txToIn), nIn(nInIn), amount(amountIn), txdata(nullptr) {}
     GenericTransactionSignatureChecker(
         const T *txToIn, unsigned int nInIn, const Amount &amountIn,
         const PrecomputedTransactionData &txdataIn)
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 9a72d0514f..af4bddc786 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -276,14 +276,17 @@ struct Stacks {
 // Extracts signatures and scripts from incomplete scriptSigs. Please do not
 // extend this, use PSBT instead
 SignatureData DataFromTransaction(const CMutableTransaction &tx,
-                                  unsigned int nIn, const CTxOut &txout) {
+                                  unsigned int nIn,
+                                  const PrecomputedTransactionData &txdata) {
     SignatureData data;
     assert(tx.vin.size() > nIn);
     data.scriptSig = tx.vin[nIn].scriptSig;
     Stacks stack(data);
+    const CTxOut &txout = txdata.m_spent_outputs[nIn];
 
     // Get signatures
-    MutableTransactionSignatureChecker tx_checker(&tx, nIn, txout.nValue);
+    MutableTransactionSignatureChecker tx_checker(&tx, nIn, txout.nValue,
+                                                  txdata);
     SignatureExtractorChecker extractor_checker(data, tx_checker);
     if (VerifyScript(data.scriptSig, txout.scriptPubKey,
                      STANDARD_SCRIPT_VERIFY_FLAGS, extractor_checker)) {
@@ -485,7 +488,7 @@ bool SignTransaction(CMutableTransaction &mtx, const SigningProvider *keystore,
         const CScript &prevPubKey = coinTxOut.scriptPubKey;
         const Amount amount = coinTxOut.nValue;
 
-        SignatureData sigdata = DataFromTransaction(mtx, i, coinTxOut);
+        SignatureData sigdata = DataFromTransaction(mtx, i, txdata);
         // Only sign SIGHASH_SINGLE if there's a corresponding output:
         if ((sigHashType.getBaseType() != BaseSigHashType::SINGLE) ||
             (i < mtx.vout.size())) {
@@ -506,7 +509,8 @@ bool SignTransaction(CMutableTransaction &mtx, const SigningProvider *keystore,
         ScriptError serror = ScriptError::OK;
         if (!VerifyScript(
                 txin.scriptSig, prevPubKey, STANDARD_SCRIPT_VERIFY_FLAGS,
-                TransactionSignatureChecker(&txConst, i, amount), &serror)) {
+                TransactionSignatureChecker(&txConst, i, amount, txdata),
+                &serror)) {
             if (serror == ScriptError::INVALID_STACK_OPERATION) {
                 // Unable to sign input and verification failed (possible
                 // attempt to partially sign).
diff --git a/src/script/sign.h b/src/script/sign.h
index a4e0e610cc..534b8d4bff 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -186,7 +186,8 @@ bool SignSignature(const SigningProvider &provider, const CTransaction &txFrom,
 
 /** Extract signature data from a transaction input, and insert it. */
 SignatureData DataFromTransaction(const CMutableTransaction &tx,
-                                  unsigned int nIn, const CTxOut &txout);
+                                  unsigned int nIn,
+                                  const PrecomputedTransactionData &txdata);
 void UpdateInput(CTxIn &input, const SignatureData &data);
 
 /**
diff --git a/src/test/fuzz/script_sign.cpp b/src/test/fuzz/script_sign.cpp
index 9cb4c5c864..b38cf82e32 100644
--- a/src/test/fuzz/script_sign.cpp
+++ b/src/test/fuzz/script_sign.cpp
@@ -99,8 +99,11 @@ FUZZ_TARGET_INIT(script_sign, initialize_script_sign) {
             fuzzed_data_provider.ConsumeIntegral<unsigned int>();
         if (mutable_transaction && tx_out &&
             mutable_transaction->vin.size() > n_in) {
+            PrecomputedTransactionData txdata(
+                *mutable_transaction,
+                std::vector<CTxOut>(mutable_transaction->vin.size(), *tx_out));
             SignatureData signature_data_1 =
-                DataFromTransaction(*mutable_transaction, n_in, {*tx_out});
+                DataFromTransaction(*mutable_transaction, n_in, txdata);
             CTxIn input;
             UpdateInput(input, signature_data_1);
             const CScript script = ConsumeScript(fuzzed_data_provider);
diff --git a/src/test/multisig_tests.cpp b/src/test/multisig_tests.cpp
index 7097051e32..f5ea447aba 100644
--- a/src/test/multisig_tests.cpp
+++ b/src/test/multisig_tests.cpp
@@ -70,11 +70,13 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
 
     // Spending transaction
     CMutableTransaction txTo[3];
+    std::vector<PrecomputedTransactionData> txdata;
     for (int i = 0; i < 3; i++) {
         txTo[i].vin.resize(1);
         txTo[i].vout.resize(1);
         txTo[i].vin[0].prevout = COutPoint(txFrom.GetId(), i);
         txTo[i].vout[0].nValue = SATOSHI;
+        txdata.push_back(PrecomputedTransactionData(txTo[i], {txFrom.vout[i]}));
     }
 
     std::vector<CKey> keys;
@@ -86,28 +88,29 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
     s = sign_multisig(a_and_b, keys, txTo[0], 0);
     BOOST_CHECK(VerifyScript(
         s, a_and_b, flags,
-        MutableTransactionSignatureChecker(&txTo[0], 0, amount), &err));
+        MutableTransactionSignatureChecker(&txTo[0], 0, amount, txdata[0]),
+        &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     for (int i = 0; i < 4; i++) {
         keys.assign(1, key[i]);
         s = sign_multisig(a_and_b, keys, txTo[0], 0);
-        BOOST_CHECK_MESSAGE(
-            !VerifyScript(
-                s, a_and_b, flags,
-                MutableTransactionSignatureChecker(&txTo[0], 0, amount), &err),
-            strprintf("a&b 1: %d", i));
+        BOOST_CHECK_MESSAGE(!VerifyScript(s, a_and_b, flags,
+                                          MutableTransactionSignatureChecker(
+                                              &txTo[0], 0, amount, txdata[0]),
+                                          &err),
+                            strprintf("a&b 1: %d", i));
         BOOST_CHECK_MESSAGE(err == ScriptError::INVALID_STACK_OPERATION,
                             ScriptErrorString(err));
 
         keys.assign(1, key[1]);
         keys.push_back(key[i]);
         s = sign_multisig(a_and_b, keys, txTo[0], 0);
-        BOOST_CHECK_MESSAGE(
-            !VerifyScript(
-                s, a_and_b, flags,
-                MutableTransactionSignatureChecker(&txTo[0], 0, amount), &err),
-            strprintf("a&b 2: %d", i));
+        BOOST_CHECK_MESSAGE(!VerifyScript(s, a_and_b, flags,
+                                          MutableTransactionSignatureChecker(
+                                              &txTo[0], 0, amount, txdata[0]),
+                                          &err),
+                            strprintf("a&b 2: %d", i));
         BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                             ScriptErrorString(err));
     }
@@ -117,18 +120,19 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
         keys.assign(1, key[i]);
         s = sign_multisig(a_or_b, keys, txTo[1], 0);
         if (i == 0 || i == 1) {
-            BOOST_CHECK_MESSAGE(VerifyScript(s, a_or_b, flags,
-                                             MutableTransactionSignatureChecker(
-                                                 &txTo[1], 0, amount),
-                                             &err),
-                                strprintf("a|b: %d", i));
+            BOOST_CHECK_MESSAGE(
+                VerifyScript(s, a_or_b, flags,
+                             MutableTransactionSignatureChecker(
+                                 &txTo[1], 0, amount, txdata[1]),
+                             &err),
+                strprintf("a|b: %d", i));
             BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
         } else {
             BOOST_CHECK_MESSAGE(
-                !VerifyScript(
-                    s, a_or_b, flags,
-                    MutableTransactionSignatureChecker(&txTo[1], 0, amount),
-                    &err),
+                !VerifyScript(s, a_or_b, flags,
+                              MutableTransactionSignatureChecker(
+                                  &txTo[1], 0, amount, txdata[1]),
+                              &err),
                 strprintf("a|b: %d", i));
             BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                                 ScriptErrorString(err));
@@ -138,7 +142,8 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
     s << OP_0 << OP_1;
     BOOST_CHECK(!VerifyScript(
         s, a_or_b, flags,
-        MutableTransactionSignatureChecker(&txTo[1], 0, amount), &err));
+        MutableTransactionSignatureChecker(&txTo[1], 0, amount, txdata[1]),
+        &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_DER, ScriptErrorString(err));
 
     for (int i = 0; i < 4; i++) {
@@ -148,19 +153,19 @@ BOOST_AUTO_TEST_CASE(multisig_verify) {
             s = sign_multisig(escrow, keys, txTo[2], 0);
             if (i < j && i < 3 && j < 3) {
                 BOOST_CHECK_MESSAGE(
-                    VerifyScript(
-                        s, escrow, flags,
-                        MutableTransactionSignatureChecker(&txTo[2], 0, amount),
-                        &err),
+                    VerifyScript(s, escrow, flags,
+                                 MutableTransactionSignatureChecker(
+                                     &txTo[2], 0, amount, txdata[2]),
+                                 &err),
                     strprintf("escrow 1: %d %d", i, j));
                 BOOST_CHECK_MESSAGE(err == ScriptError::OK,
                                     ScriptErrorString(err));
             } else {
                 BOOST_CHECK_MESSAGE(
-                    !VerifyScript(
-                        s, escrow, flags,
-                        MutableTransactionSignatureChecker(&txTo[2], 0, amount),
-                        &err),
+                    !VerifyScript(s, escrow, flags,
+                                  MutableTransactionSignatureChecker(
+                                      &txTo[2], 0, amount, txdata[2]),
+                                  &err),
                     strprintf("escrow 2: %d %d", i, j));
                 BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                                     ScriptErrorString(err));
diff --git a/src/test/script_p2sh_tests.cpp b/src/test/script_p2sh_tests.cpp
index 8b431a76fe..c3c8dd717a 100644
--- a/src/test/script_p2sh_tests.cpp
+++ b/src/test/script_p2sh_tests.cpp
@@ -44,11 +44,11 @@ static bool Verify(const CScript &scriptSig, const CScript &scriptPubKey,
     txTo.vin[0].scriptSig = scriptSig;
     txTo.vout[0].nValue = SATOSHI;
 
-    return VerifyScript(
-        scriptSig, scriptPubKey,
-        SCRIPT_ENABLE_SIGHASH_FORKID,
-        MutableTransactionSignatureChecker(&txTo, 0, txFrom.vout[0].nValue),
-        &err);
+    PrecomputedTransactionData txdata(txTo, std::vector(txTo.vout));
+    return VerifyScript(scriptSig, scriptPubKey, SCRIPT_ENABLE_SIGHASH_FORKID,
+                        MutableTransactionSignatureChecker(
+                            &txTo, 0, txFrom.vout[0].nValue, txdata),
+                        &err);
 }
 
 BOOST_FIXTURE_TEST_SUITE(script_p2sh_tests, BasicTestingSetup)
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index a1811e49b3..7c301af3d8 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -132,11 +132,13 @@ struct ScriptTest : BasicTestingSetup {
             BuildCreditingTransaction(scriptPubKey, nValue)};
         const CMutableTransaction tx =
             BuildSpendingTransaction(scriptSig, txCredit);
-        BOOST_CHECK_MESSAGE(VerifyScript(scriptSig, scriptPubKey, flags,
-                                         MutableTransactionSignatureChecker(
-                                             &tx, 0, txCredit.vout[0].nValue),
-                                         &err) == expect,
-                            message);
+        const PrecomputedTransactionData txdata(tx, std::vector(txCredit.vout));
+        BOOST_CHECK_MESSAGE(
+            VerifyScript(scriptSig, scriptPubKey, flags,
+                         MutableTransactionSignatureChecker(
+                             &tx, 0, txCredit.vout[0].nValue, txdata),
+                         &err) == expect,
+            message);
         BOOST_CHECK_MESSAGE(err == scriptError,
                             FormatScriptError(err) + " where " +
                                 FormatScriptError(scriptError) +
@@ -158,7 +160,7 @@ struct ScriptTest : BasicTestingSetup {
             BOOST_CHECK_MESSAGE(
                 VerifyScript(scriptSig, scriptPubKey, combined_flags,
                              MutableTransactionSignatureChecker(
-                                 &tx, 0, txCredit.vout[0].nValue),
+                                 &tx, 0, txCredit.vout[0].nValue, txdata),
                              &err) == expect,
                 message + strprintf(" (with %s flags %08x)",
                                     expect ? "removed" : "added",
@@ -2203,35 +2205,36 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG12) {
     const CTransaction txFrom12{
         BuildCreditingTransaction(scriptPubKey12, Amount::zero())};
     CMutableTransaction txTo12 = BuildSpendingTransaction(CScript(), txFrom12);
+    PrecomputedTransactionData txdata(txTo12, std::vector(txFrom12.vout));
 
     CScript goodsig1 =
         sign_multisig(scriptPubKey12, key1, CTransaction(txTo12));
-    BOOST_CHECK(VerifyScript(
-        goodsig1, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig1, scriptPubKey12, gFlags,
+                             MutableTransactionSignatureChecker(
+                                 &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
     txTo12.vout[0].nValue = 2 * SATOSHI;
-    BOOST_CHECK(!VerifyScript(
-        goodsig1, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(goodsig1, scriptPubKey12, gFlags,
+                              MutableTransactionSignatureChecker(
+                                  &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
     CScript goodsig2 =
         sign_multisig(scriptPubKey12, key2, CTransaction(txTo12));
-    BOOST_CHECK(VerifyScript(
-        goodsig2, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig2, scriptPubKey12, gFlags,
+                             MutableTransactionSignatureChecker(
+                                 &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     CScript badsig1 = sign_multisig(scriptPubKey12, key3, CTransaction(txTo12));
-    BOOST_CHECK(!VerifyScript(
-        badsig1, scriptPubKey12, gFlags,
-        MutableTransactionSignatureChecker(&txTo12, 0, txFrom12.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig1, scriptPubKey12, gFlags,
+                              MutableTransactionSignatureChecker(
+                                  &txTo12, 0, txFrom12.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 }
@@ -2259,45 +2262,46 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     // we can convert it to readonly transaction and use
     // TransactionSignatureChecker instead of MutableTransactionSignatureChecker
     const CTransaction txTo23(mutableTxTo23);
+    const PrecomputedTransactionData txdata(txTo23, std::vector(txFrom23.vout));
 
     std::vector<CKey> keys;
     keys.push_back(key1);
     keys.push_back(key2);
     CScript goodsig1 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(VerifyScript(
-        goodsig1, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig1, scriptPubKey23, gFlags,
+                             TransactionSignatureChecker(
+                                 &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     keys.clear();
     keys.push_back(key1);
     keys.push_back(key3);
     CScript goodsig2 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(VerifyScript(
-        goodsig2, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig2, scriptPubKey23, gFlags,
+                             TransactionSignatureChecker(
+                                 &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     keys.clear();
     keys.push_back(key2);
     keys.push_back(key3);
     CScript goodsig3 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(VerifyScript(
-        goodsig3, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(VerifyScript(goodsig3, scriptPubKey23, gFlags,
+                             TransactionSignatureChecker(
+                                 &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                             &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
 
     keys.clear();
     keys.push_back(key2);
     keys.push_back(key2); // Can't re-use sig
     CScript badsig1 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig1, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig1, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2305,10 +2309,10 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key2);
     keys.push_back(key1); // sigs must be in correct order
     CScript badsig2 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig2, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig2, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2316,10 +2320,10 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key3);
     keys.push_back(key2); // sigs must be in correct order
     CScript badsig3 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig3, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig3, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2327,10 +2331,10 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key4);
     keys.push_back(key2); // sigs must match pubkeys
     CScript badsig4 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig4, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig4, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
@@ -2338,19 +2342,19 @@ BOOST_AUTO_TEST_CASE(script_CHECKMULTISIG23) {
     keys.push_back(key1);
     keys.push_back(key4); // sigs must match pubkeys
     CScript badsig5 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig5, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig5, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::SIG_NULLFAIL,
                         ScriptErrorString(err));
 
     keys.clear(); // Must have signatures
     CScript badsig6 = sign_multisig(scriptPubKey23, keys, txTo23);
-    BOOST_CHECK(!VerifyScript(
-        badsig6, scriptPubKey23, gFlags,
-        TransactionSignatureChecker(&txTo23, 0, txFrom23.vout[0].nValue),
-        &err));
+    BOOST_CHECK(!VerifyScript(badsig6, scriptPubKey23, gFlags,
+                              TransactionSignatureChecker(
+                                  &txTo23, 0, txFrom23.vout[0].nValue, txdata),
+                              &err));
     BOOST_CHECK_MESSAGE(err == ScriptError::INVALID_STACK_OPERATION,
                         ScriptErrorString(err));
 }
@@ -2400,7 +2404,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     // Single signature case:
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined = CombineSignatures(txFrom.vout[0], txTo, scriptSig, empty);
     BOOST_CHECK(combined.scriptSig == scriptSig.scriptSig);
     combined = CombineSignatures(txFrom.vout[0], txTo, empty, scriptSig);
@@ -2409,7 +2414,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     // Signing again will give a different, valid signature:
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined =
         CombineSignatures(txFrom.vout[0], txTo, scriptSigCopy, scriptSig);
     BOOST_CHECK(combined.scriptSig == scriptSigCopy.scriptSig ||
@@ -2423,7 +2429,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     txTo.vin[0].prevout = COutPoint(txFrom.GetId(), 0);
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined = CombineSignatures(txFrom.vout[0], txTo, scriptSig, empty);
     BOOST_CHECK(combined.scriptSig == scriptSig.scriptSig);
     combined = CombineSignatures(txFrom.vout[0], txTo, empty, scriptSig);
@@ -2431,7 +2438,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     scriptSigCopy = scriptSig;
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined =
         CombineSignatures(txFrom.vout[0], txTo, scriptSigCopy, scriptSig);
     BOOST_CHECK(combined.scriptSig == scriptSigCopy.scriptSig ||
@@ -2443,7 +2451,8 @@ BOOST_AUTO_TEST_CASE(script_combineSigs) {
     BOOST_CHECK(keystore.AddCScript(scriptPubKey));
     BOOST_CHECK(SignSignature(keystore, CTransaction(txFrom), txTo, 0,
                               SigHashType().withForkId()));
-    scriptSig = DataFromTransaction(txTo, 0, txFrom.vout[0]);
+    scriptSig = DataFromTransaction(
+        txTo, 0, PrecomputedTransactionData(txTo, std::vector(txFrom.vout)));
     combined = CombineSignatures(txFrom.vout[0], txTo, scriptSig, empty);
     BOOST_CHECK(combined.scriptSig == scriptSig.scriptSig);
     combined = CombineSignatures(txFrom.vout[0], txTo, empty, scriptSig);
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index 17387a993e..afbd2c6928 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -356,10 +356,11 @@ static void CheckWithFlag(const CTransactionRef &output,
                           bool success) {
     ScriptError error;
     CTransaction inputi(input);
+    const PrecomputedTransactionData txdata(inputi, {output->vout[0]});
     bool ret = VerifyScript(
         inputi.vin[0].scriptSig, output->vout[0].scriptPubKey,
         flags | SCRIPT_ENABLE_SIGHASH_FORKID,
-        TransactionSignatureChecker(&inputi, 0, output->vout[0].nValue),
+        TransactionSignatureChecker(&inputi, 0, output->vout[0].nValue, txdata),
         &error);
     BOOST_CHECK_EQUAL(ret, success);
 }
@@ -477,8 +478,10 @@ SignatureData CombineSignatures(const CMutableTransaction &input1,
                                 const CMutableTransaction &input2,
                                 const CTransactionRef tx) {
     SignatureData sigdata;
-    sigdata = DataFromTransaction(input1, 0, tx->vout[0]);
-    sigdata.MergeSignatureData(DataFromTransaction(input2, 0, tx->vout[0]));
+    sigdata = DataFromTransaction(
+        input1, 0, PrecomputedTransactionData(input1, {tx->vout[0]}));
+    sigdata.MergeSignatureData(DataFromTransaction(
+        input2, 0, PrecomputedTransactionData(input2, {tx->vout[0]})));
     const PrecomputedTransactionData txdata(*tx, {tx->vout[0]});
     ProduceSignature(DUMMY_SIGNING_PROVIDER,
                      MutableTransactionSignatureCreator(
